packages/connect-node/src/private/node-universal.ts

```diff
@@ -0,0 +1,173 @@
+// Copyright 2021-2023 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import type * as http from "http";
+import type * as http2 from "http2";
+import { write } from "./io.js";
+import type {
+  UniversalHandlerFn,
+  UniversalRequest,
+  UniversalResponse,
+} from "../protocol-handler.js";
+
+/**
+ * NodeHandler is compatible with http.RequestListener and its equivalent
+ * for http2.
+ */
+export type NodeHandler = (
+  request: NodeRequest,
+  response: NodeResponse
+) => void;
+type NodeRequest = http.IncomingMessage | http2.Http2ServerRequest;
+type NodeResponse = http.ServerResponse | http2.Http2ServerResponse;
+
+/**
+ * Convert a universal handler to a Node.js handler function.
+ */
+export function universalHandlerToNodeHandler(
```

Comment:
For handlers, this is now the single bridge between Node.js and our implementation.